### PR TITLE
11700 user table refactor deletion

### PR DESCRIPTION
**Pull request by javitonino**
**REMEMBER TO REMOVE RAISES (revert last commit) BEFORE DEPLOY**

#11700 

With this, all tests in `table_spec.rb` pass. Next step would be to change the `model_class` to `Carto::UserTable` and see what breaks. I made a [quick experiment](https://ci.cartodb.net/job/CartoDB-PR-testing-backend/79/console) to facilitate writing the next issue.

Implemented in this PR:
- Table deletion hooks, including visualization model deletion (and friends)
- Table blender compatibility with both models (this one changes production code that is being actively used, but should be safe)
- Junk deletion. Notably `UserTable.search` which is not being used in production code (now searches go through VQB).
- The typical bunch of test fixes. I have not reorganized `table_spec.rb` again, so we can see changes more clearly.

**Acceptance**
- [x] Ghost tables. Rename a table via SQL API, refresh the dashboard. Ghost tables should fix it.
- [x] Ghost tables. Create a table via SQL API (and cartodbfy) it, refresh the dashboard. Ghost tables should create metadata.
- [x] Ghost tables. Drop a table via SQL API, refresh the dashboard. Ghost tables should delete metadata.
- [x] Table blender. Create a map from one dataset (in dashboard). Check privacy of created map.
- [x] Table blender. Create a map from multiple dataset (in dashboard). Check privacy of created map.
- [x] Table blender. Create a map from dataset (from dataset view). Check privacy of created map.
- [x] Delete a map.
- [x] Delete a dataset (with maps depending on it, partially and fully).
- [x] Set a new custom basemap. Delete the map. Create a new map. The previous custom basemap should be one of the options for custom basemaps.
- [x] Publish a public map. Open the embed and the vizjson. Change privacy to private and republish. Reload the embed and vizjson. Both should throw an error.
- [x] Create a map, open the vizjson. Delete the map and refresh the vizjson. It should throw an error.

### CODE REVIEW

**Comment by javitonino** [comment]
retest this please

**Comment by javitonino** [comment]
retest this, please

**Comment by juanignaciosl** [comment]
Additional acceptance suggestions (WIP):
- [ ] Set a new custom basemap. Delete the map. Create a new map. The previous custom basemap should be one of the options for custom basemaps.

File: app/models/carto/user_table.rb [hunk 1]
```diff
@@ -51,6 +51,8 @@ def self.column_defaults
     after_create :create_canonical_visualization
     after_create { service.after_create }
     after_save { CartoDB::Logger.debug(message: "Carto::UserTable#after_save"); service.after_save }
+    before_destroy { service.before_destroy unless destroyed? }
```
Comment (juanignaciosl) [inline app/models/carto/user_table.rb]: Is this `destroyed?` check needed to avoid double triggering? Maybe a comment might be useful.

File: app/models/map.rb [hunk 2]
```diff
@@ -200,6 +200,10 @@ def view_bounds_data
     }
   end
 
+  def dup
```
Comment (juanignaciosl) [inline app/models/map.rb]: How come `Layer` can just alias `copy` but `Map` needs this?

File: app/models/table/privacy_manager.rb [hunk 3]
```diff
@@ -73,8 +73,10 @@ def propagate_to(visualizations, table_privacy_changed = false)
         if visualization.type == CartoDB::Visualization::Member::TYPE_CANONICAL
           # Each table has a canonical visualization which must have privacy synced
           visualization.privacy = ::UserTable::PRIVACY_VALUES_TO_TEXTS[privacy]
+          visualization.store_using_table(table_privacy_changed)
```
Comment (juanignaciosl) [inline app/models/table/privacy_manager.rb]: Why is this change needed?

File: app/models/carto/visualization.rb [hunk 4]
```diff
@@ -96,6 +96,10 @@ class Carto::Visualization < ActiveRecord::Base
   after_save :save_named_map_or_rollback_privacy, :propagate_attribution_change
   after_save :propagate_privacy_and_name_to, if: :table
 
+  before_destroy :backup_visualization
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: I think that we should follow the same approach than we did in the previous PR, and log usages. It was really useful. What do you think?

PS: for acceptance, I'd raise instead of logging in those cases that are not legit.

File: app/models/map.rb (hunk 2 above)
Comment (javitonino) [inline app/models/map.rb]: `copy` is a custom method in `Layer`, there is no `copy` in Sequel. I created this and copied the implementation from the map copier.

File: app/models/carto/user_table.rb (hunk 1 above)
Comment (javitonino) [inline app/models/carto/user_table.rb]: Yes, it avoids triggering deletion hooks twice. It is caused like this:
1. `::Table.destroy`
2. `::Table.after_save`
3. `Visualization.destroy`
4. Following `dependent: :destroy`, it deletes map, and the UserTable
5. `Carto::UserTable.before/after_destroy`

It was one of the first changes (minimum to get tests to pass), and looking at it now, I don't like it. The thing is, `after_destroy` is still being triggered twice (and adding the `unless destroyed?` doesn't work there, as it would never trigger). I think we need to improve this in order to not trigger `after_save` twice, although, right now it's safe.

I may be overthinking this, and maybe the best course of action is to simply document this, so we can be careful in the future. I think there might be a an opportunity to clean this up when deletion the `::UserTable` model, by moving much of the `before/after_destroy` logic to the model (the metadata part).

🤔 I'm rubber-ducking a bit, but I think a TODO comment might be enough. Ok for you?

File: app/models/table/privacy_manager.rb (hunk 3 above)
Comment (javitonino) [inline app/models/table/privacy_manager.rb]: Basically, what this loops wants to do is:
- For canonical visualizations, change the privacy and invalidate cache
- For derived visualizations, invalidate the cache

The old code relied on an implementation detail in `store` where `table_privacy_changed = true` forced the invalidation to trigger. In the new model, the decision to invalidate or not is based on checking if attributes have changed.

I think the changed code makes this behaviour more explicit.

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: Yup, some logging would be nice. I think the number of code paths is smaller here, but I'll have a pass at adding them.

File: app/models/carto/user_table.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/models/carto/user_table.rb]: According to your explanation, there's no easy way to fix this for both models. If that's true, document this and we'll address this once there's only one model to worry about 👍

File: app/models/carto/visualization.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Add it, then :-)

**Comment by juanignaciosl** [comment]
👍  after adding logging.

**Comment by javitonino** [comment]
Added logging, -> to acceptance. Adding suggestions to top list.

**Comment by juanignaciosl** [comment]
👍 , removing raises...
